Edit block style variations from global styles

tellthemachines opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Closes #44417.

This is an initial approach to editing block style variations in global styles. It's still a work in progress and not the complete implementation for #44417.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Currently this PR only allows to edit existing variations, which works to a limited extent, given some of the classic core block variations style elements not yet available as block supports, or aren't added to every block. It also leverages the existing global styles interface for now, so the "variations" section appears as if it were another block support. 

#### Limitations

* Style variations can only make use of whatever the block supports, so their scope is much more limited than the classic CSS variations.

#### Todoes

* It should be possible to add new style variations to any block;
* "Variations" should be a section of its own, as shown in the designs in #44417.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. In Global Styles, navigate to "Blocks" and then into a block that has variations, e.g. Button, Image, or Site Logo.
2. Check that a "Variations" section is rendered inside the Block screen and that clicking on that section brings up the Variations screen;
3. Check that inside the Variations screen all core variations for the block are rendered;
4. Clicking into the variations, it should be possible to change their styles and those changes should only affect that specific variation.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

Global Styles can be accessed from the Site Editor header by clicking the "Styles" button. Unfortunately keyboard navigation to the sidebar is broken, so it is necessary to tab 5-6 times after clicking "Styles" to actually arrive in the sidebar. (The first button in the sidebar is called "Open Style Book"; from then on tabbing a few more times should take us to the "Blocks" section)

## Screenshots or screencast <!-- if applicable -->

<img width="275" alt="Screenshot 2022-12-20 at 2 19 08 pm" src="https://user-images.githubusercontent.com/8096000/208573637-3359b07c-1bed-48f7-9cce-1b883ad35755.png">
<img width="281" alt="Screenshot 2022-12-20 at 2 19 37 pm" src="https://user-images.githubusercontent.com/8096000/208573646-cd199ab5-24c3-4413-a708-300035fc0b3f.png">
<img width="282" alt="Screenshot 2022-12-20 at 2 19 46 pm" src="https://user-images.githubusercontent.com/8096000/208573654-0db6744e-940b-4bc4-91ff-cefaed9b89d4.png">

tellthemachines committed:
Edit block style variations from global styles

tellthemachines committed:
It's not working yet but we'll get there

tellthemachines committed:
Hardcoded a bunch of stuff to see if it works

tellthemachines committed:
Add back empty line

tellthemachines committed:
update panel paths to use encodeURIComponent

tellthemachines committed:
Move updated sanitize function to 6.2 folder

tellthemachines committed:
Output CSS for edited style variations

tellthemachines committed:
Any [a-z]* style variation name should be possible

tellthemachines committed:
Fix test failures

tellthemachines committed:
Add handling of feature selectors

tellthemachines committed:
Add colors and typography

tellthemachines committed:
Re-add changes to class-wp-theme-json

tellthemachines committed:
Fix color paths

tellthemachines committed:
Add dimensions styles

tellthemachines committed:
Bump up front end specificity.

tellthemachines committed:
Only core variations should be editable

jameskoster commented:
Would it be possible to include the block preview (with the style variation applied) here?

<img width="279" alt="Screenshot 2022-12-20 at 11 23 40" src="https://user-images.githubusercontent.com/846565/208655967-a031454a-f2fa-4e6e-ab3c-0d5ef437e038.png">

I guess the variation title should be capitalised as well.

Not something to fix here, but it's very easy to get lost in all the Styles panels, and style variations are going to add another layer of complexity. For example the screenshot above – you have nothing to really orient where you are. It would be good to explore options.

jasmussen commented:
Nice, this is really close already. Here's a GIF showing the global styles interface of changing the button block "fill" variation:

![variations](https://user-images.githubusercontent.com/1204802/208670190-a4b97c32-f8e6-441f-965d-ddb2c5a962e6.gif)

Will it be possible to get closer to the ItemGroup based visuals of the original design?

ramonjd commented:
Testing this. Looking great so far. 

- Changes I make to block variations in the site editor are saved and are reflected when I select the corresponding block variations in the editor 👍 

I had some observations while testing. Some may not be related to the changes in this PR:

1. This is a question of design, but I'm wondering if the interface should resemble the block styles we see in the post editor (with the buttons and preview). Reason being, is that "variation" is an internal term so I trying to find a connect between the two.
2. Potential bug: in the site editor go to styles > color > text/background and try to select a text color. The sidebar goes blank. Same is for all color panels, e.g.,  insert a table in the site editor,  colors > text or block > table > variations > default > colors > text. I'm not sure, but maybe something to do with navigation rendering? 
3. The image and table styles aren't reflected in the site editor. I think this might be due to the way the styles are applied using `__experimentalSelector` ? To replicate, insert a table or image block in the site editor, then modify the block's variation styles. 
4. This is probably not a huge deal: the default variation global styles don't apply to existing blocks. So, create a post with a table block but don't click on a style variation button (the default variation is "Default"). Save the post then go to the site editor, and set some styles for the table block default variation. Save, then go back to the previous post and refresh the page. The styles aren't applied, and only show up when I toggle variations. This is because the style variation classes aren't added by default, `is-style-regular`. Not sure how we get around that unless we apply the style class to blocks by default, and risk a deprecation hailstorm 🤔 
5. The ToolsPanel reset button doesn't seem to have any effect. E.g., change the border values of the image block variations in the site editor. Save, then return and try to reset them. I think this is related to https://github.com/WordPress/gutenberg/issues/46505

tellthemachines commented:
Thanks for all the feedback folks!

> Would it be possible to include the block preview (with the style variation applied) here?

Sure!

> Will it be possible to get closer to the ItemGroup based visuals of the original design?

Absolutely, this first pass is more aiming to get the functionality right, and I'll look into putting the actual UI in place as per the designs next 😄 

> Potential bug: in the site editor go to styles > color > text/background and try to select a text color. The sidebar goes blank

Actual bug 😅  I've been messing around with the navigation paths to integrate the variations screens and must have broken something, thanks for spotting it!

> The image and table styles aren't reflected in the site editor. I think this might be due to the way the styles are applied using __experimentalSelector ? To replicate, insert a table or image block in the site editor, then modify the block's variation styles.

Hmm, I can't reproduce this unless with the default state as you report in your 4th point. Could it be the same issue?

tellthemachines committed:
Basic functionality to create a new variation

ramonjd commented:
> Hmm, I can't reproduce this unless with the default state as you report in your 4th point. Could it be the same issue?

Ah yeah, sorry for the red herring. 

I retested and you're right: is when the default variation class isn't applied (on a freshly-added block)

tellthemachines commented:
Update: I just pushed the beginnings of adding new style variations. Please ignore what it looks like for now 😅 as I just dumped a button in there to be able to test the functionality. So far it only works for blocks that already have variations, though it should be easy to expand it to all blocks if we want.

What _doesn't_ work at all is getting the newly added variations to show in the post editor. Unless I'm mistaken, we're going to need something like #34178 in place before we can do that.

My feeling is that the ability to add new variations is the key piece here, as opposed to editing existing ones. Would you say that @jasmussen , @jameskoster ? If so, it might make sense to put this on hold until we get #34178 done.

In any case, I'll shortly be off for the holidays so don't worry if this PR isn't updated for the next week or so, I haven't forgotten it!

jasmussen commented:
> My feeling is that the ability to add new variations is the key piece here, as opposed to editing existing ones. 

I think that is a key part of it, absolutely. Though depending on what's technically feasible, I would suggest it equally important to edit and customize the existing ones, especially if such changes can reflect on existing content without needing an update. I.e. if I change the default button "outline" style to have a dashed border, would that automatically translate to every existing piece on the site using outline buttons?

Enjoy your AFK, thanks for all the work here!

tellthemachines commented:
> I change the default button "outline" style to have a dashed border, would that automatically translate to every existing piece on the site using outline buttons?

Yeah, that already works in this PR, except in the case @ramonjd pointed out above: if the block hasn't been changed from its initial default style, it won't have the style classname attached. We'll have to look into how best to approach this. Adding the default style classname to every block would land us in deprecation hell 😅 

We could try to add the class only if the default style has been changed. I don't think inline styles are a great option here because these are global, so we want to keep their specificity fairly low.

If editing existing styles is useful by itself, then I'll leave out adding new ones for now, get existing ones working well, and then tackle the new ones separately after #34178 is done.

jasmussen commented:
> Adding the default style classname to every block would land us in deprecation hell 😅

Is it possible to treat the default style in a special manner, both with regards to the block (class or not), and with regards to the variations interface? By simply knowing what the default is, we might be able to make these aspects controllable, maybe? Apologies if I'm oversimplifying! 🙏

tellthemachines committed:
Revert adding new variations for now

tellthemachines committed:
Add preview panel to variation screen.

tellthemachines committed:
ItemGroup UI

tellthemachines committed:
Fix color screen bug

tellthemachines committed:
Disable editing on default styles.

tellthemachines commented:
> Is it possible to treat the default style in a special manner, both with regards to the block (class or not), and with regards to the variations interface? By simply knowing what the default is, we might be able to make these aspects controllable, maybe?

I've updated this PR to disable editing the default style. I'm not sure if that's what you had in mind 😅 but it might be enough for a V1.

If we do want the default to also be editable, we'll have to look into dynamically adding the classname to all blocks that have edits to that variation. This should be doable in the front end, but it's a bit more tricky in the editor as we'll need something like #34178 in place to know if variations have been edited.

I've also changed the UI to match the designs on #44417. The main I'm missing is the icon in front of each variation. Is it meant to be a tiny preview? Do we have a component for it or will it have to be created from scratch?

The bug with the colors screens not rendering has also been fixed.

I've removed the button to add a new variation as we'll have to get #34178 done before tackling that.

apeatling commented:
Coming into this fresh and editing a style variation made sense to me and was clear. Nice work!

A few things that I noticed without any previous context in my head:

* I expected to be able to add a style variation, so looking forward to trying that out.
* I didn't understand why I couldn't click and edit the default variation, it seemed broken to me (no explanation or difference in the UI other than it being unresponsive).
* If I go out of the Button view, and go back in, no variation is selected but the main preview still shows the last selected variation, not the default. I would expect the variation to be selected to match what is shown in the main preview? See here:

![2023-01-04 10 12 40](https://user-images.githubusercontent.com/1464705/210621821-fa2311c2-e4a7-4ef4-b9c6-41001c790131.gif)

tellthemachines committed:
Fix variation showing on general block preview

tellthemachines commented:
Thanks for testing @apeatling ! I've fixed the bug with preview showing the last selected variation.

> I didn't understand why I couldn't click and edit the default variation, it seemed broken to me (no explanation or difference in the UI other than it being unresponsive).

Yeah, it isn't great; I was hoping for some design input on that one. The problem is that if we do edit the default variation, there's currently no way of making it apply to blocks because they don't have the default variation classname explicitly applied. @jasmussen mentioned treating the default specially above,but I think I need a little more guidance on what we should be aiming for there 😅 

> I expected to be able to add a style variation, so looking forward to trying that out.

That will be part 2 of this work, because it depends on #34178 being done first.

jameskoster commented:
> The main I'm missing is the icon in front of each variation. Is it meant to be a tiny preview?

Yup, it's the same as the ones you find in the Typography section.

---

I noticed an issue with the preview thumbnail, it seems to display the default variation when editing different properties of non-default variations:

https://user-images.githubusercontent.com/846565/210802278-c83b321b-d5e4-4f45-90b3-426dd768ca36.mp4

---

As for editing the default variation... I agree this is confusing. That said, isn't the default 'variation' just the default block styles, IE it isn't actually a variation at all? If that's the case, might it be best not to treat it as a variation in this UI?

tellthemachines commented:
> As for editing the default variation... I agree this is confusing. That said, isn't the default 'variation' just the default block styles, IE it isn't actually a variation at all? If that's the case, might it be best not to treat it as a variation in this UI?

That's a good point! Removing the default altogether from this UI would definitely make things easier. It'll still show as a style variation in the block settings though - might that be confusing for users? Should we somehow flag it as default in the block settings?

tellthemachines committed:
Fix previews for design tools screens inside variations.

andrewserong commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -713,9 +713,17 @@ protected static function sanitize( $input, $valid_block_names, $valid_element_n
-			$schema_settings_blocks[ $block ]           = static::VALID_SETTINGS;
-			$schema_styles_blocks[ $block ]             = $styles_non_top_level;
-			$schema_styles_blocks[ $block ]['elements'] = $schema_styles_elements;
+			// Blocks can have multiple style variations, so we need to build the schema for each one.
```
Tiny nit: during previous backporting I've gotten feedback to avoid using "we" in comments. Would something like the following work?

```suggestion
			// Build the schema for each block style variation.
```

tellthemachines replied:
Oh yes that sounds much better, thanks!

andrewserong commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -863,6 +871,15 @@ protected static function get_blocks_metadata() {
+			if ( ! empty( $block_type->styles ) ) {
+				$style_selectors = array();
+				foreach ( $block_type->styles as $style ) {
+					$style_selectors[ $style['name'] ] = static::append_to_selector( '.is-style-' . $style['name'] . '.is-style-' . $style['name'], static::$blocks_metadata[ $block_name ]['selector'] );
```
Just double checking — this adds the classname twice, which increases specificity. Is this to ensure that styles included in `theme.json` override the defaults for the style variation provided in CSS stylesheets? If so, that sounds like a decent way to ensure it overrides them to me! Is it worth adding a comment prior to this line explaining the extra specificity?

tellthemachines replied:
Yeah, I duplicated the classname to make sure these styles override the block core styles. To answer your question above, I didn't do the same in the editor because`editor-styles-wrapper` already guarantees the needed specificity. It's a good idea to add a comment here though!

andrewserong commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -2230,6 +2258,47 @@ public function get_styles_for_block( $block_metadata ) {
 			}
 		}
 
+		// If there are style variations, generate the selectors for them, including any feature selectors.
```
It looks like this code block generates the whole CSS declarations, and not just the selectors. Should we update the comment to reflect that?

andrewserong commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -2230,6 +2258,47 @@ public function get_styles_for_block( $block_metadata ) {
+				$style_variation_node     = _wp_array_get( $this->theme_json, $style_variation['path'], array() );
+				$style_variation_selector = $style_variation['selector'];
+
+				if ( ! empty( $block_metadata['features'] ) ) {
```
Would it be worth adding a comment prior to this line that this section processes feature level selectors for each style variation?

tellthemachines replied:
It's always worth adding comments! Also I need to get better at remembering that 😅

andrewserong commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -2230,6 +2258,47 @@ public function get_styles_for_block( $block_metadata ) {
+						}
+					}
+				}
+				$style_variation_declarations[ $style_variation_selector ] = static::compute_style_properties( $style_variation_node, $settings, null, $this->theme_json );
```
It took me a second to work out why this area has two calls to `compute_style_properties`. It looks like the second call here is for anything not already covered by feature level selectors (since the feature level ones will have unset the feature from the `$style_variation_node`).

Would it be worth adding a comment prior to this line, something along the lines of "Compute declarations for remaining styles not covered by feature level selectors" or something like that?

andrewserong commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -2230,6 +2258,47 @@ public function get_styles_for_block( $block_metadata ) {
+							} else {
+								$style_variation_declarations[ $combined_feature_selectors ] = $new_feature_declarations;
+							}
+							// Remove the feature from the block's node now the
```
Tiny nit: this is updating the variation node, not the block node:

```suggestion
							// Remove the feature from the variation's node now the
```

tellthemachines replied:
Well spotted, I copy-pasted from above 😂

andrewserong commented on packages/edit-site/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -673,6 +687,83 @@ export const toStyles = (
+											if (
+												!! featureDeclarations.length
+											) {
+												const _featureSelectors =
```
Tiny nit: Are the underscore prefixes used here to avoid collisions with the `featureSelectors` variable? For readability, would it be worth splitting out any of this logic into a separate function since the nesting here is fairly deep?

tellthemachines replied:
I'll try and take this bit out to see if it improves readability.

andrewserong commented on packages/edit-site/src/components/global-styles/variations-panel.js:
```diff
@@ -0,0 +1,95 @@
+			{ coreBlockStyles.map( ( style, index ) => {
+				if ( style?.isDefault ) {
+					return (
+						<NavigationButtonAsItem
```
For the default item, would it be worth making sure it visually looks as though it's either disabled, or not a button? I'm not quite sure how we'd flag it in the UI, but from a user perspective, I imagine we might want to somehow communicate that for the default style, simply adjust the styles as you usually would — i.e. you only need to dive into variations if you're adjusting the non-default styles? (Please correct me if I'm wrong about that assumption!)

Alternately, as others mentioned, maybe it'd be easier to remove the default option from here — but as you say, that could also be confusing for a user if it isn't there at all 🤔

tellthemachines replied:
I might try and remove the default for now and see how it goes! It's definitely the easier option.

andrewserong reviewed (commented):
Great work on all this @tellthemachines! I was about to comment on the block previews within each of the different property screens, but you fixed that up while I was reviewing 😄

I left a few comments, mostly nits / code quality questions about the PHP bits — feel free to ignore any you don't think are relevant!

Overall it's testing really nicely for me, and I like that for now it just focuses on the core style variations, so that block styles bundled with themes aren't included 👍

In terms of the CSS output, I noticed that the specificity is slightly different between the style output in global styles versus on the site front end. It looks like on the site front end, the style classname is duplicated — I was wondering if that's intentional to ensure greater specificity, and whether that's also needed in global styles? Although, in global styles, we already have the `.editor-styles-wrapper ` prefix, so maybe it's a little different anyway 🤔

| In global styles | On site front end (note double style classname) |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/210907692-cd5016ba-47c8-413a-b5ab-f817a2e32a41.png) | ![image](https://user-images.githubusercontent.com/14988353/210907711-2e8edfc3-db2c-43c1-9110-427ed9975365.png) |

This PR looks really close to land to me — it sounds like one of the last things to figure out at this stage is whether to remove the Default variation button?

tellthemachines committed:
Address feedback.

tellthemachines committed:
Remove default variation from panel.

tellthemachines commented:
Thanks for all the feedback, folks!

@jameskoster regarding the Typography icons, those are static SVGs for which we only change the colors when selected. If we want the icons to preview the style variations, we'll need something like the block preview mechanism, using example markup from each block and adding the classname for each preview. The problem is with scaling block previews down to icon size: though we could conceivably change the example markup to fit a small size for e.g. Button block which is mainly text, that won't work so well for Image because we have to take caption into account too. Might we be able to do without an icon here? Or perhaps we could look into changing the top preview on hover/focus on each variation?

jameskoster commented:
> It'll still show as a style variation in the block settings though - might that be confusing for users? Should we somehow flag it as default in the block settings?

Like this?

<img width="297" alt="Screenshot 2023-01-06 at 12 10 45" src="https://user-images.githubusercontent.com/846565/211010346-1c475fb7-120a-41f2-b9c3-327a2cac1752.png">

It seems fine to me.

apeatling commented:
I would +1 not showing default in this list, especially because the initial view is editing the default already. It's like you're already in default before you select a variation.

tellthemachines commented:
I've removed the default styles from the list and addressed all the code-related feedback. The only pending issue to address is the icons. As I mentioned [here](https://github.com/WordPress/gutenberg/pull/46343#issuecomment-1373107836) it's not staightforward to make them behave as style previews, so I'd appreciate some design feedback on this. Cc @jameskoster @jasmussen

jameskoster commented:
That works well, thanks. We should probably update the UI here too, right?

<img width="314" alt="Screenshot 2023-01-10 at 11 33 22" src="https://user-images.githubusercontent.com/846565/211541465-d7c8d726-a51b-4986-82c7-4d320f31738c.png">

The default "variation" (which isn't actually a variation) should be presented as "Default".

---

I don't know if it needs to happen in this PR, but we need to do something about orientation... There's nothing here that clarifies you're editing the outline variation of the button block. The preview helps, but it's not conclusive – you might have two similar variations.

<img width="281" alt="Screenshot 2023-01-10 at 11 33 07" src="https://user-images.githubusercontent.com/846565/211541627-80c731fd-c881-424d-a641-26114656254a.png">

The UI to edit the full color palette also feels way too prominent in this context.

tellthemachines committed:
Default style should always be called Default.

tellthemachines commented:
> The default "variation" (which isn't actually a variation) should be presented as "Default".

I've added some logic to make sure the default variation is always called "Default" in the block settings UI.

> I don't know if it needs to happen in this PR, but we need to do something about orientation... There's nothing here that clarifies you're editing the outline variation of the button block. The preview helps, but it's not conclusive – you might have two similar variations.

Agreed! I think this would be best addressed as a follow-up given it seems to need some design work.

I guess this one's ready for a final review! Tests are all passing now so 🤞 will stay that way.

andrewserong commented on packages/edit-site/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -892,12 +1005,19 @@ export function useGlobalStylesOutput() {
 		}
 		mergedConfig = updateConfigWithSeparator( mergedConfig );
-		const blockSelectors = getBlockSelectors( getBlockTypes() );
+		const blockSelectors = getBlockSelectors(
+			getBlockTypes(),
+			getBlockStyles
+		);
```
Not a blocker, but in a code quality follow-up, would it be worth switching `getBlockSelectors` over to be a hook instead so we don't need to pass it the `getBlockStyles` function?

andrewserong commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -2230,6 +2259,49 @@ public function get_styles_for_block( $block_metadata ) {
 			}
 		}
 
+		// If there are style variations, generate the declarations for them, including any feature selectors the block may have.
```
Not a blocker, but I wonder if in a follow-up there's an opportunity to either consolidate some of this logic with the feature declarations above, or split it out into a separate function, since this function is now quite long? Definitely not something to worry about in this PR, just jotting down the thought 🙂

andrewserong reviewed (approved):
Just gave this a re-test and it's all still working nicely for me 👍

| In site editor | In post editor |
| --- | --- |
| <img width="275" alt="image" src="https://user-images.githubusercontent.com/14988353/211726163-3a3aea91-9a6d-42d8-aa32-1293f4a9f38f.png"> | <img width="167" alt="image" src="https://user-images.githubusercontent.com/14988353/211726150-af95743e-64ed-4430-a5d3-8ae8b03eb669.png"> |

Just left a couple of comments about potential code quality follow-ups, we might also want to add or update the theme JSON PHP tests to include coverage for dealing with style variations? It'll probably help once we get to the backporting process, but I don't think it's a blocker for now since we've been manually testing it pretty thoroughly so far.

> I've added some logic to make sure the default variation is always called "Default" in the block settings UI.

This sounds fine to me, I suppose an argument could be made that the block's code that registers the block style could make the call on what the label should say (the separator block's default style is called "Default", for example). However, since users can change the default style to whatever they like via global styles, it's highly possible that the name for the default style might not make sense depending on the settings used... e.g. "Fill" may not really be "Fill" for the Button block. It'd be easy to change back, though, so a +1 from me for going with hard-coding it to "Default" and we can see if there's any feedback on it.

Overall this looks in good shape to me, I reckon it's in a good place to land! ✨

tellthemachines commented:
Thanks for reviewing @andrewserong ! Those are great suggestions and I'll follow up with another PR shortly.

Yeah, I thought about just changing the name of the button style to Default, but went with this to make sure it always displays as "Default" so people know it's the default style for that block.

jameskoster commented:
Nice work on this one 🥳 

> Agreed! I think this would be best addressed as a follow-up given it seems to need some design work.

I opened a couple of issues: https://github.com/WordPress/gutenberg/issues/47021, https://github.com/WordPress/gutenberg/issues/47026.

fabiankaegy commented:
@bph I realize it is probably already a bit late but I think this here should really get a dev note / some documentation for the 6.2 release

bph commented:
@fabiankaegy thank you so much for flagging this. It's not too late :-)

tellthemachines commented:
Here goes the dev note:

## Edit block style variations from theme.json

This new theme.json API allows existing core block style variations to be styled directly from the `theme.json` file. It does not currently allow adding new variations. Styles should be added as follows:

```
	"styles": {
		"blocks": {
			"core/image": {
				"variations": {
					"rounded": {
						"border": {
							"width": "10px",
							"color": "#cc3ee6"
						}	
					}
				}
			}
                }
       }
```

Each variation should be targeted by its name, which can be found in the block's `block.json` file, under "styles".

There is also a corresponding UI in the block section of global styles:

<img width="280" alt="Screenshot 2023-03-03 at 10 23 55 am" src="https://user-images.githubusercontent.com/8096000/222588258-ae2d3f53-7435-4a79-90ef-13a5433fae42.png">

oandregal commented:
:wave: Prepared a follow-up at https://github.com/WordPress/gutenberg/pull/49807

I've also noticed that there are no docs for this at https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/ It'd be great to have some.

oandregal commented:
PR to add docs for `variations` in `theme.json` at https://github.com/WordPress/gutenberg/pull/49826
